### PR TITLE
Introduce baseline JIT compiler

### PR DESCRIPTION
**Pull request by qwe661234**
The baseline JIT compiler design involves tracing an EBB when its usage frequency exceeds a predetermined threshold. We then utilize a code generator to convert the instruction sequence into C code. The C code is subsequently compiled using mir, and the resulting target machine code is stored in the code cache for future utilization.

The primary objective of introducing the baseline JIT compiler is to enhance the execution speed of RISC-V instructions with minimal modification. This implementation requires three additional components: a code generator, mir compiler, and code cache, while maintaining the original design of the interpreter. Furthermore, this baseline JIT compiler serves as the foundational target for future improvements.

In addition, we have developed a Python script that effectively traces code templates and automatically generates JIT code templates. This approach eliminates the need for manually writing duplicated codey.

* Core i7-8700

| Metric    |     master      |  baseline JIT   |Speedup |
|-----------|-----------------|-----------------|--------|
| CoreMark  |1352.843 (iter/s)|1932.496 (iter/s)| +42.8% |
| Dhrystone |1282 DMIPS       |2495.11 DMIPS    | +94.6% |

* eMAG

| Metric    |     master      |  baseline JIT   |Speedup |
|-----------|-----------------|----------------|--------|
| CoreMark  | 438.073 (iter/s)| 644.426 (iter/s)| +47.1% |
| Dhrystone | 354.60 DMIPS    | 867.00 DMIPS    |  +144% |

Known issue: this implementation does not correctly work on Apple M1 MacOS.

### CODE REVIEW

**Review by jserv — CHANGES_REQUESTED** [review]
You shall rework the commit "Decouple RV32 instructions as implementation templates" and then rebase. i.e., at least two commits.

**Review by jserv — CHANGES_REQUESTED** [review]
Explicitly mention commit 54024f74d7f1b86bddae0b011bd0469e1ba378db while comparing with JIT compilation in git commit message.

File: src/emulate.c [hunk 1]
```diff
@@ -310,971 +171,10 @@
...
-     * is misaligned or if the memory chunk does not exist.
-     */
-    RV_EXC_MISALIGN_HANDLER(3, load, false, 1);
-    rv->X[fuse[0].rd] = rv->io.mem_read_w(rv, addr);
-    for (int i = 1; i < ir->imm2; i++) {
-        addr = rv->X[fuse[i].rs1] + fuse[i].imm;
-        rv->X[fuse[i].rd] = rv->io.mem_read_w(rv, addr);
-    }
-})
+/* FIXME: Add JIT-based execution path */
```
[the remark this answers is not in the document]
Comment (jserv) [reply src/emulate.c]: Remove this comment.

File: src/rv32_template.c [hunk 2]
```diff
@@ -0,0 +1,946 @@
...
+RVOP(fcvtwus, { rv->X[ir->rd] = (uint32_t) rv->F[ir->rs1]; })
+
+/* FMV.X.W */
+RVOP(fmvxw, { rv->X[ir->rd] = rv->F_int[ir->rs1]; })
+
+/* FEQ.S performs a quiet comparison: it only sets the invalid operation
+ * exception flag if either input is a signaling NaN.
+ */
+RVOP(feqs, {
+    rv->X[ir->rd] = (rv->F[ir->rs1] == rv->F[ir->rs2]) ? 1 : 0;
```
[the remark this answers is not in the document]
Comment (jserv) [reply src/rv32_template.c:658]: @qwe661234, let's create another pull request for fixing this.

File: src/cache.h [hunk 3]
```diff
@@ -37,4 +38,8 @@ void *cache_put(struct cache *cache, uint32_t key, void *value);
  * @cache: a pointer points to target cache
  * @callback: a function for freeing cache entry completely
  */
-void cache_free(struct cache *cache, void (*callback)(void *));
\ No newline at end of file
+void cache_free(struct cache *cache, void (*callback)(void *));
+
+#if RV32_HAS(JIT)
+bool cache_hot(struct cache *cache, uint32_t key);
```
Comment (jserv) [inline src/cache.h:50]: Add comments here.

File: src/compile.c [hunk 4]
```diff
@@ -0,0 +1,411 @@
...
+    uint8_t debug_level;
+    uint8_t optimize_level;
+} riscv_jit_t;
+
+typedef struct {
+    char *name;
+    void *func;
+} func_obj_t;
+
+#define SET_SIZE_BITS 10
```
Comment (jserv) [inline src/compile.c:40]: Add some comments here, illustrating set data structure.

File: src/compile.c [hunk 5]
```diff
@@ -0,0 +1,411 @@
...
+        if (!strcmp(name, func_list[i].name))
+            return func_list[i].func;
+    }
+    return NULL;
+}
+
+static riscv_jit_t *jit = NULL;
+static jit_item_t *jit_ptr = NULL;
+
+/* TODO: fix error when running on ARM architecture */
```
Comment (jserv) [inline src/compile.c]: You shall describe the problem statement in details.

File: src/compile.c [hunk 6]
```diff
@@ -0,0 +1,430 @@
...
+#include "mir-gen.h"
+#include "mir.h"
+#include "riscv_private.h"
+#include "utils.h"
+
+typedef struct {
+    char *code;
+    size_t code_size;
+    size_t curr;
+} jit_item_t;
```
Comment (jserv) [inline src/compile.c]: I am afraid that `jit_item_t` is a bit misleading. Did you mean that it is a matter of code cache?

Comment (qwe661234) [reply src/compile.c]: This is a data structure used by c2mir to retrieve the code string.

Comment (qwe661234) [reply src/compile.c]: Maybe we can rename it to `jit_gencode_t`?

Comment (jserv) [reply src/compile.c]: How about `code_string_t`? By the way, you shall provide more relevant comments.

File: src/riscv_private.h [hunk 7]
```diff
@@ -12,7 +12,9 @@
 #endif
 #include "decode.h"
 #include "riscv.h"
-
+#if RV32_HAS(JIT)
+#include "cache.h"
+#endif
```
Comment (jserv) [inline src/riscv_private.h:17]: Append a new line right after `#endif`.

File: src/riscv_private.h [hunk 8]
```diff
@@ -119,8 +126,13 @@ struct riscv_internal {
     uint32_t csr_mip;
     uint32_t csr_mbadaddr;
 
-    bool compressed;       /**< current instruction is compressed or not */
+    bool compressed; /**< current instruction is compressed or not */
+#if !RV32_HAS(JIT)
     block_map_t block_map; /**< basic block map */
+#else
+    struct cache *cache;
```
Comment (jserv) [inline src/riscv_private.h]: The naming of this object `cache` is misleading. Can you clarify?

File: src/riscv_private.h [hunk 9]
```diff
@@ -59,8 +61,12 @@ typedef struct block {
     uint32_t insn_capacity;    /**< maximum of instructions encompased */
     struct block *predict;     /**< block prediction */
     rv_insn_t *ir;             /**< IR as memory blocks */
+#if RV32_HAS(JIT)
+    bool hot; /**< Determine the block is hotspot or not */
+#endif
 } block_t;
 
+#if !RV32_HAS(JIT)
```
Comment (jserv) [inline src/riscv_private.h]: Drop this preprocessor directive since it is a well-defined declaration. Keeping it doesn't hurt.

Comment (jserv) [reply src/riscv_private.h]: Add comments to annotate the functionality of this struct.

File: src/riscv_private.h [hunk 10]
```diff
@@ -59,8 +62,13 @@ typedef struct block {
     uint32_t insn_capacity;    /**< maximum of instructions encompased */
     struct block *predict;     /**< block prediction */
     rv_insn_t *ir;             /**< IR as memory blocks */
+#if RV32_HAS(JIT)
+    bool hot; /**< Determine the block is hotspot or not */
+#endif
 } block_t;
 
+/* map the translated basic block to its program counter  to accelerate
+ * searching. */
```
Comment (jserv) [inline src/riscv_private.h]: Mention that this struct is used only in interpreter at the moment.

File: src/compile.c (hunk 4 above)
Comment (jserv) [reply src/compile.c:40]: ?

File: src/compile.c [hunk 11]
```diff
@@ -0,0 +1,435 @@
...
+    strcat(gencode, PROLOGUE);
+    set_reset(&set);
+    block_t *block = cache_get(rv->block_cache, rv->PC);
+    trace_ebb(rv, gencode, block->ir, &set);
+    strcat(gencode, EPILOGUE);
+}
+
+static int get_string_func(void *data)
+{
+    code_string_t *jit_code_string = data;
```
Comment (jserv) [inline src/compile.c]: Replace `jit_code_string` with `codestr` for short.

File: src/compile.c [hunk 12]
```diff
@@ -0,0 +1,435 @@
...
+#include "mir-gen.h"
+#include "mir.h"
+#include "riscv_private.h"
+#include "utils.h"
+
+/* Provide for c2mir to retrieve the generated code string. */
+typedef struct {
+    char *code;
+    size_t code_size;
+    size_t curr;
```
Comment (jserv) [inline src/compile.c]: You shall explain the functionality of `curr`.

File: src/compile.c [hunk 13]
```diff
@@ -0,0 +1,435 @@
...
+
+#define SET_SIZE_BITS 10
+#define SET_SIZE 1 << SET_SIZE_BITS
+#define HASH(val) \
+    (((val) * (GOLDEN_RATIO_32)) >> (32 - SET_SIZE_BITS)) & ((SET_SIZE) -1)
+#define sys_icache_invalidate(addr, size) \
+    __builtin___clear_cache((char *) (addr), (char *) (addr) + (size));
+
+/*
+ *  The set consists of SET_SIZE buckets, with each bucket containing 32 slots.
```
Comment (jserv) [inline src/compile.c]: Why 32 slots?

File: src/compile.c [hunk 14]
```diff
@@ -0,0 +1,435 @@
...
+}
+
+/**
+ * set_has - check whether the element exist in the set or not
+ * @set: a pointer points to target set
+ * @key: the key of the inserted entry
+ */
+static bool set_has(set_t *set, uint32_t key)
+{
+    uint32_t index = HASH(key);
```
Comment (jserv) [inline src/compile.c]: Add `const` qualifier for `index`.

File: src/compile.c [hunk 15]
```diff
@@ -0,0 +1,435 @@
...
+    uint8_t count = 0;
+    while (set->table[index][count]) {
```
Comment (jserv) [inline src/compile.c]: Rewrite as following:
```c
    for (uint8_t count = 0; set->table[index][count]; count++) {
```

File: src/compile.c [hunk 16]
```diff
@@ -0,0 +1,435 @@
...
+    if (!jit) {
+        jit = calloc(1, sizeof(riscv_jit_t));
+        jit->options = calloc(1, sizeof(struct c2mir_options));
+        jit->ctx = MIR_init();
+        jit->optimize_level = 1;
+    }
+
+    if (!jit_code_string) {
+        jit_code_string = malloc(sizeof(code_string_t));
+        jit_code_string->curr = 0;
```
Comment (jserv) [inline src/compile.c]: Move the statement `jit_code_string->curr = 0;` out of if-else block.

File: src/compile.c (hunk 13 above)
Comment (qwe661234) [reply src/compile.c]: I just set a sufficiently large number to prevent the bucket from becoming full.

File: src/compile.c (hunk 4 above)
Comment (qwe661234) [reply src/compile.c:40]: I add comments of set structure below.

File: src/compile.c (hunk 13 above)
Comment (jserv) [reply src/compile.c]: Add more comments and mark it as configurable. We shall always distinguish between magic numbers and configurations.

File: src/emulate.c [hunk 17]
```diff
@@ -1313,6 +211,7 @@ static inline bool insn_is_unconditional_branch(uint8_t opcode)
     return false;
 }
 
+#if !RV32_HAS(JIT)
```
Comment (jserv) [inline src/emulate.c:215]: Is it possible to unify hash function implementation? i.e., unify macro `HASH` and function `hash`.

Comment (qwe661234) [reply src/emulate.c:215]: It is possible, but it may impact the performance of interpreter.

Comment (jserv) [reply src/emulate.c:215]: Add FIXME/TODO comments within this pull request. Then, submit an issue on GitHub to address code consolidation.

**Pull request merged**